### DISCUSSION TITLE
The Future of Sorting/Filtering

### DISCUSSION THREAD

**Discussion by exformation** [ppy/osu#27555] [Ideas]
There's some overlap here with previous discussions but I wanted to consolidate and append. 

Generally, it'd be nice to open up more of the data associated with maps to the user, so they can sort and filter by whatever they might think of. These additions might end up being so extensive that rethinking the UI for sorting/filtering might be needed, so that users aren't presented with 50 item lists. I've experimented with implementing some of these but didn't want to get too deep without feedback. 

# Extending Existing Code 
Current Sorting Methods (SortMode.cs, CarouselBeatmapSet.cs, ...):
- Artist, Author, BPM, DateSubmitted, DateAdded, DateRanked, LastPlayed, Difficulty, Length, Source, Title  

Potential Additions:
- Plays, Rating, Favorites (from BM listing) 
- Accuracy, Combo, Rank, Score (highest achieved by user on that map)
- UserPlayCount
- OD, HP, AR, CS, SilderCount, CircleCount, SpinnerCount, MaxCombo, keys (some ruleset-specific)
---
Current Filtering Methods (FilterQueryParser.cs, FilterCriteria.cs, ...):
- StarDifficulty, ApproachRate, DrainRate, CircleSize, OverallDifficulty, Length, BPM, BeatDivisor, OnlineStatus, LastPlayed, Creator, Artist, Title, DifficultyName, keys (mania) 

Potential Additions:
- all suggested sort methods
  - ex: acc<95 fav>100 plays<10 maxCombo>2000
- FeaturedArtist, Category (ranked, qualified, etc.) Genre, Language, HasStoryboard, HasVideo (from BM listing)
  - ex: genre=metal cat=qualified fa=true video=true

# Bigger Changes
## Sorting/Filtering By Pattern
Process maps for basic patterns, like how many doubles, triples, quads, and quints they have, average pattern length, average stream length, average even/odd pattern count, etc.
These could also be mode specific, like wanting to sort/filter by TrillCount, StairCount, ChordCount, etc. for mania. 

ex: "I want to practice maps that have lots of doubles and quads"
-> sortBy evenPatternCount
or
-> evenPatternCount>oddPatternCount

## Custom Sorting/Filtering
Allow for some logic in defining sorting/filtering methods. 

ex: "I want to play rhythmically dense stuff with a variety of pattern lengths"
-> sortBy min(doubleCount,tripleCount,quadCount,quintCount)/circleCount

ex: "I want to practice chords"
-> sortBy chordCount/circleCount

ex: "I want to play deathstreams"
-> median(patternLength)>20

# Notes
Some of these might seem weird, like allowing users to sort by HP instead of just filtering for what HP they want, but I figure if there's something a map *can* be sorted/filtered by, it might as well be opened up to the user. It could allow them to do some niche things, like sortBy hp+od. 

I realize people might not like the option for custom logic: <https://github.com/ppy/osu/discussions/26582#discussioncomment-8173704> 

I constantly find myself wanting to do programmatic filtering/sorting like shown above so that I can find maps that are relevant to me. It's fun. External tools can do this, but it'd be nice if they populated the actual song select. If stuff like this isn't added to the game, is it okay to create tools that can override the maps listed in song select of a running client? It'd require messing with RAM values I'm guessing.

**Comment by Rekunan** [comment]
> Sorting/Filtering By Pattern

This whole section isn't even possible yet as it'd all be new calculations, which I think at least some should be put in the map details, as I've suggested here: #27262

**Comment by respektive** [comment]
I would definitely love to filter by individual object counts and max combo as well as the total object count. I think for sorting, having all of them might be a bit excessive? I would probably only add `TotalObjectCount` and `MaxCombo` to the sorting options. Though my only reasoning for that would be that the sorting methods are already kinda annoying to use and adding even more would just make it worse. But the new song select redesign gets rid of the list of sorting methods and only has a single drop down to choose from, which you should be able to search in, it would probably be fine to add more sorting methods then.

Also something to note is that most of these sorting methods (and even some of the current ones like bpm or length) would need split out individual difficulties instead of sorting the sets. The way it's currently implemented for difficulty sort, with each difficulty still showing the beatmapset card, is not really that great imo.

tl;dr: Basically what I'm trying to say is that this would probably be nice additions for when the new song select has been implemented I guess. lol

**Comment by TextAdventurer12** [comment]
it would also be nice for sorting to apply after mods instead of before as they currently do

**Reply by Rekunan** [reply]
Luckily, there are tracking issues for this:
#12628
#19947

**Comment by Wilker-uwu** [comment]
not sure if i should comment here but it would also be cool to have some form of tag wrangling of the same kind that ArchiveOfOurOwn does. a decision process of "should a `skillset/jumpstream` tag marked by players also imply the `skillset/streams` also needs to be marked automatically?", or "should a search for `lastplayed` imply `played=yes`?", or "should `drumandbass` equal `dnb` and imply `electronic`?"
